File: code/ch13/python/feature_flags.py
# §13.3.3 Feature Flags — a release flag, then the same check as a percentage rollout
# (canary deployment's progressive exposure implemented in code).
# Runs standalone: python3 feature_flags.py
from zlib import crc32

def render_new(user_id): return f"new:{user_id}"      # stand-ins for the real pages
def render_old(user_id): return f"old:{user_id}"

def scheduler_page(user_id, flags):
  if flags["new_scheduler"]:                        # release flag: one bit, everyone
    return render_new(user_id)
  return render_old(user_id)

def scheduler_page_rollout(user_id, flags):           # same conditional, new predicate
  if crc32(user_id.encode()) % 100 < flags["new_scheduler_pct"]:  # stable bucket 0..99
    return render_new(user_id)
  return render_old(user_id)

if __name__ == "__main__":
  users = [f"user{i}" for i in range(10_000)]
  assert all(scheduler_page(u, {"new_scheduler": True}).startswith("new") for u in users)
  assert all(scheduler_page(u, {"new_scheduler": False}).startswith("old") for u in users)
  assert all(scheduler_page_rollout(u, {"new_scheduler_pct": 0}).startswith("old")
       for u in users)
  assert all(scheduler_page_rollout(u, {"new_scheduler_pct": 100}).startswith("new")
       for u in users)
  hits = sum(scheduler_page_rollout(u, {"new_scheduler_pct": 3}).startswith("new")
       for u in users)
  print(f"3% rollout reached {hits} of {len(users)} users ({hits / len(users):.1%})")
  # the same user always lands in the same bucket, across runs and processes
  assert (scheduler_page_rollout("user42", {"new_scheduler_pct": 3})
      == scheduler_page_rollout("user42", {"new_scheduler_pct": 3}))
  print("all assertions passed")

File: code/ch13/java/FeatureFlags.java
// §13.3.3 Feature Flags — a release flag, then the same check as a percentage rollout,
// Java variant. String.hashCode() is fixed by the Java spec, so buckets are stable
// across JVMs — but they differ from the CRC-32 buckets of the Python and Ruby
// variants (any stable hash works, but pick one per system).
// Run: java -ea FeatureFlags.java
public class FeatureFlags {
  record Flags(boolean newScheduler, int newSchedulerPct) {}

  static String renderNew(String userId) { return "new:" + userId; } // stand-ins for
  static String renderOld(String userId) { return "old:" + userId; } // the real pages

  static String schedulerPage(String userId, Flags flags) {
    if (flags.newScheduler()) {                     // release flag: one bit, everyone
      return renderNew(userId);
    }
    return renderOld(userId);
  }

  static String schedulerPageRollout(String userId, Flags flags) { // same conditional
    // String.hashCode() is spec-fixed, so buckets are stable across JVMs — but they
    // differ from the CRC-32 buckets the Python and Ruby variants compute
    if (Math.floorMod(userId.hashCode(), 100) < flags.newSchedulerPct()) { // 0..99
      return renderNew(userId);
    }
    return renderOld(userId);
  }

  public static void main(String[] args) {
    int hits = 0;
    for (int i = 0; i < 10_000; i++) {
      String u = "user" + i;
      assert schedulerPage(u, new Flags(true, 0)).startsWith("new");
      assert schedulerPage(u, new Flags(false, 0)).startsWith("old");
      assert schedulerPageRollout(u, new Flags(false, 0)).startsWith("old");
      assert schedulerPageRollout(u, new Flags(false, 100)).startsWith("new");
      if (schedulerPageRollout(u, new Flags(false, 3)).startsWith("new")) hits++;
    }
    System.out.printf("3%% rollout reached %d of 10000 users (%.1f%%)%n",
        hits, hits / 100.0);
    // the same user always lands in the same bucket, across runs and processes
    assert schedulerPageRollout("user42", new Flags(false, 3))
        .equals(schedulerPageRollout("user42", new Flags(false, 3)));
    System.out.println("all assertions passed");
  }
}
